### PR TITLE
Add timing output to the end of ngen runs

### PR DESCRIPTION
**Pull request by PhilMiller**
This will allow for slightly more informative benchmark results, in which we can distinguish different elements of the runtime, and extrapolate accordingly.

## Changes

- Simplify the mixed MPI/Routing `#ifdef` logic
- Add time calls and prints

## Testing

1. Unit testing still passes

## Screenshot

From the CI job running on example data:

```
Running timestep 700
Finished 720 timesteps.
NGen top-level timings:
	NGen::init: 0.207376
	NGen::simulation: 0.230584
	NGen::routing: 1.232e-06
```

## Todos

- Improve numeric formatting

## Checklist

- [x] PR has an informative and human-readable title
- [x] Changes are limited to a single goal (no scope creep)
- [x] Code can be automatically merged (no conflicts)
- [x] Code follows project standards (link if applicable)
- [x] Passes all existing automated tests
- [x] Any _change_ in functionality is tested
- [x] New functions are documented (with a description, list of inputs, and expected output)
- [x] Placeholder code is flagged / future todos are captured in comments
- [x] Project documentation has been updated (including the "Unreleased" section of the CHANGELOG)
- [x] Reviewers requested with the [Reviewers tool](https://help.github.com/articles/requesting-a-pull-request-review/) :arrow_right:

### Target Environment support

- [x] Linux
- [x] macOS

### CODE REVIEW

File: src/NGen.cpp [hunk 1]
```diff
@@ -51,7 +52,9 @@ bool is_subdivided_hydrofabric_wanted = false;
 std::string PARTITION_PATH = "";
 int mpi_rank;
```
Comment (hellkite500) [inline src/NGen.cpp:53]:
```suggestion

```

File: src/NGen.cpp [hunk 2]
```diff
@@ -51,7 +52,9 @@ bool is_subdivided_hydrofabric_wanted = false;
...
 int mpi_num_procs;
-#endif
+#else // NGEN_MPI_ACTIVE
+int mpi_rank = 0;
```
Comment (hellkite500) [inline src/NGen.cpp:56]:
```suggestion

```

File: src/NGen.cpp [hunk 3]
```diff
@@ -51,7 +52,9 @@ bool is_subdivided_hydrofabric_wanted = false;
...
+#endif // NGEN_MPI_ACTIVE
```
Comment (hellkite500) [inline src/NGen.cpp:58]: if mpi_rank is used regardless of the macro value, declare & define once outside the macro logic.
```suggestion
int mpi_rank = 0;
```

File: src/NGen.cpp [hunk 4]
```diff
@@ -557,13 +573,31 @@ int main(int argc, char *argv[]) {
...
     }
- #endif //NGEN_MPI_ACTIVE
- #else
- #ifdef NGEN_MPI_ACTIVE
+#endif
+
+    auto time_done_routing = std::chrono::steady_clock::now();
+    auto time_elapsed_routing = time_done_routing - time_done_simulation;
+
+    auto as_seconds = [](decltype(time_elapsed_routing) in)
```
Comment (hellkite500) [inline src/NGen.cpp:582]: move this lambda inside the `if(mpi_rank == 0)`?  Only rank 0 is using it.

File: src/NGen.cpp (hunk 3 above)
Comment (PhilMiller) [reply src/NGen.cpp:58]: Sure

File: src/NGen.cpp (hunk 4 above)
Comment (PhilMiller) [reply src/NGen.cpp:582]: Sure. I expect it'll eventually end up as a standalone function somewhere in a utility header, but I can keep it super-local for now until it has more usage.

**Comment by PhilMiller** [comment]
Ugh, and of course in tweaking things between my last local build and pushing, I broke it a bit. That'll be fixed shortly.

**Comment by ZacharyWills** [comment]
This will greatly help with hardware affinity testing. Ideally we should be able to give some rough numbers about the relationship between domain extent, available hardware, and expected runtime.

Thanks,
Zach

**Comment by robertbartel** [comment]
I'll leave it to others to formally approve, but I like it overall.  Thanks for putting this together @PhilMiller.

**Comment by PhilMiller** [comment]
Ideally, the textual numeric output would be presented a little more readably, using something like https://github.com/martinmoene/EngFormat-Cpp but that's an inessential refinement, compared to just getting meaningful measurements in place.

All previous comments on the code addressed.

**Review by robertbartel — APPROVED** [review]

**Comment by ZacharyWills** [comment]
@PhilMiller 
Just for clarity, what units are the timings in?

```
	NGen::init: 0.207376
	NGen::simulation: 0.230584
	NGen::routing: 1.232e-06
```

**Comment by PhilMiller** [comment]
@ZacharyWills Seconds

**Pull request merged**